### ISSUE TITLE
Use Atom Services API to support beautifier providers

### ISSUE THREAD

**Issue by patrick-steele-idem** [Glavin001/atom-beautify#1007]
I really appreciate all of the work that has gone into maintaining `atom-beautify`, but I feel like the different beautifiers belong in their own packages. This would reduce the amount of work put on you and it would allow each beautifier to have its own release lifecycle and tests.

Have you considered using the Atom Services API to separate out beautification providers?:

- [Atom blog: New Services API](http://blog.atom.io/2015/03/25/new-services-API.html)
- [Atom API Docs: Interacting With Other Packages Via Services](http://flight-manual.atom.io/behind-atom/sections/interacting-with-other-packages-via-services/)

After working on a new provider for [autocomplete-plus](https://github.com/atom/autocomplete-plus) I feel like this would be a great fit for this project as well.

Just wanted to get your thoughts. Thanks for your time!

**Comment by Glavin001** [comment]
I've been thinking about this quite a lot and I do think moving towards Atom Services API with providers is the way to go for official version 1.0.0 of Atom-Beautify.

Right now, Atom-Beautify's core is changing often. Even keeping up with changes to Atom's API is sometimes a problem when they break things. There are times when I need to break things in Atom-Beautify, too, and I manually go through each of the languages and/or beautifier files and fix them up and then make sure the tests pass again. If these were separate packages, it would become much more time consuming for myself to maintain individually, even with the help of awesome contributors.

I want to keep understanding the needs of the users and third-party beautifiers and once we have a very strong foundation then we can start breaking it down into Services / Providers. Once we all discuss and agree on what is needed and design the final version, I want to build the final beautifier core and break things up into cohesive pieces.

This is an older comment that you can read below **What needs to happen* to see some of what I have planned: https://github.com/Glavin001/atom-beautify/issues/620#issuecomment-151024633
Unfortunately, I need time to work on designing and building this out. So I ask for everyone's patience in the mean time.

Thank you for your interest and feedback!

**Comment by patrick-steele-idem** [comment]
Sorry for missing the earlier question, but glad to see you are already thinking about this. How about we pick one beautifier to split out and just use that to nail down the API? I'll volunteer `atom-beautifier-marko` :) Also, you can still create a super package that bundles the third-party package as a `package-dep` in the main package's `package.json` (e.g. [go-plug/package.json/](https://github.com/joefitzgerald/go-plus/blob/master/package.json#L43-L53)). We can do this before `1.0.0` and the change should not impact any existing users.

For Marko, specifically, it makes sense for the provider to include both the "language definition" and the "beautifier". It also probably makes sense for each provider to have its own settings pane UI.

Longer term, you might want to decide what are "core" beautifiers that are always included and which ones are optional.

Also, with the Atom service providers implementation it looks like you can handle breaking changes between versions.

If you want to set up some time to discuss please let me know. I'll be happy to help out as best I can.

**Comment by Glavin001** [comment]
I am investigating this now. Since I am on vacation for the next couple weeks there may not be any published changes, however I'll be doing research into what Atom-Beautify v1 should look like and taking notes to implement later. Thank you for your patience.

Closing this issue in favour of https://github.com/Glavin001/atom-beautify/issues/620#issuecomment-151024633 which I explained in more detail what features I wanted to implement for v1.
